File: backend/models/content_output.py
from __future__ import annotations

import uuid
from datetime import datetime

from sqlalchemy import String, Integer, Text, ForeignKey, Enum as SAEnum, func
from sqlalchemy.dialects.postgresql import UUID, JSONB
from sqlalchemy.orm import Mapped, mapped_column, relationship

from backend.database import Base


class ContentOutput(Base):
    __tablename__ = "content_outputs"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    competitor_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("competitors.id"), nullable=False
    )
    content_type: Mapped[str] = mapped_column(String, nullable=False)
    title: Mapped[str | None] = mapped_column(String, nullable=True)
    content: Mapped[str] = mapped_column(Text, nullable=False)
    source_card_ids: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    version: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    status: Mapped[str] = mapped_column(
        SAEnum("draft", "generating", "in_review", "approved", "published", "failed", name="content_output_status"),
        default="draft",
        nullable=False,
    )
    template_id: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("content_templates.id"), nullable=True
    )
    google_doc_id: Mapped[str | None] = mapped_column(String, nullable=True)
    google_doc_url: Mapped[str | None] = mapped_column(String, nullable=True)
    approved_by: Mapped[uuid.UUID | None] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id"), nullable=True
    )
    approved_at: Mapped[datetime | None] = mapped_column(nullable=True)
    published_at: Mapped[datetime | None] = mapped_column(nullable=True)
    raw_llm_output: Mapped[dict | None] = mapped_column(JSONB, nullable=True)
    error_message: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(default=func.now(), nullable=False)
    updated_at: Mapped[datetime] = mapped_column(default=func.now(), onupdate=func.now(), nullable=False)

    competitor = relationship("Competitor")
    template = relationship("ContentTemplate")
    approver = relationship("User")

File: frontend/src/types/content-output.ts
export type ContentOutputStatus =
  | "draft"
  | "generating"
  | "in_review"
  | "approved"
  | "published"
  | "failed";

export interface ContentOutput {
  id: string;
  competitor_id: string;
  content_type: string;
  title: string;
  content: string;
  template_id: string | null;
  source_card_ids: string[];
  version: number;
  status: ContentOutputStatus;
  google_doc_id: string | null;
  google_doc_url: string | null;
  approved_by: string | null;
  approved_by_name: string | null;
  approved_at: string | null;
  published_at: string | null;
  raw_llm_output: Record<string, unknown> | null;
  error_message: string | null;
  competitor_name: string | null;
  created_at: string;
  updated_at: string;
}

